hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/ProtoConversionUtil.java

```diff
@@ -0,0 +1,393 @@
+          return tmpValue;
+        case MAP:
+          Map<Object, Object> mapValue = (Map) value;
+          Map<Object, Object> mapCopy = new HashMap<>(mapValue.size());
+          for (Map.Entry<Object, Object> entry : mapValue.entrySet()) {
+            mapCopy.put(convertObject(STRINGS, entry.getKey()), convertObject(schema.getValueType(), entry.getValue()));
+          }
+          return mapCopy;
+        case NULL:
+          return null;
+        case RECORD:
+          GenericData.Record newRecord = new GenericData.Record(schema);
+          Message messageValue = (Message) value;
+          for (Schema.Field f : schema.getFields()) {
+            int position = f.pos();
+            Descriptors.FieldDescriptor fieldDescriptor = getOrderedFields(schema, messageValue)[position];
+            Object convertedValue;
+            if (fieldDescriptor.getType() == Descriptors.FieldDescriptor.Type.MESSAGE && !fieldDescriptor.isRepeated() && !messageValue.hasField(fieldDescriptor)) {
+              convertedValue = null;
+            } else {
+              convertedValue = convertObject(f.schema(), messageValue.getField(fieldDescriptor));
+            }
+            newRecord.put(position, convertedValue);
+          }
+          return newRecord;
+        case STRING:
+          if (value instanceof String) {
+            return value;
+          } else if (value instanceof StringValue) {
+            return ((StringValue) value).getValue();
+          } else {
+            return new Utf8(value.toString());
+          }
+        case UNION:
+          // Unions only occur for nullable fields when working with proto + avro and null is the first schema in the union
+          return convertObject(schema.getTypes().get(1), value);
+        default:
+          throw new HoodieException("Proto to Avro conversion failed for schema \"" + schema + "\" and value \"" + value + "\"");
+      }
+    }
+
+    /**
+     * Returns the wrapped field, assumes all wrapped fields have a single value
+     * @param value wrapper message like {@link Int32Value} or {@link StringValue}
+     * @return the wrapped object
+     */
+    private Object getWrappedValue(Object value) {
+      Message valueAsMessage = (Message) value;
+      return valueAsMessage.getField(valueAsMessage.getDescriptorForType().getFields().get(0));
+    }
+
+    private String getNamespace(Descriptors.FileDescriptor fd, Descriptors.Descriptor containing) {
+      DescriptorProtos.FileOptions filedOptions = fd.getOptions();
+      String classPackage = filedOptions.hasJavaPackage() ? filedOptions.getJavaPackage() : fd.getPackage();
+      String outer = "";
+      if (!filedOptions.getJavaMultipleFiles()) {
+        if (filedOptions.hasJavaOuterClassname()) {
+          outer = filedOptions.getJavaOuterClassname();
+        } else {
+          outer = new File(fd.getName()).getName();
+          outer = outer.substring(0, outer.lastIndexOf('.'));
+          outer = toCamelCase(outer);
```

Comment:
why do we need to do this? is it a convention?

Reply:
I simplified this namespace generation to just use the proto package